### PR TITLE
[SPARK-26002][SQL] Fix day of year calculation for Julian calendar days

### PR DESCRIPTION
**Pull request by attilapiros**
## What changes were proposed in this pull request?

Fixing leap year calculations for date operators (year/month/dayOfYear) where the Julian calendars are used (before 1582-10-04). In a Julian calendar every years which are multiples of 4 are leap years (there is no extra exception for years multiples of 100).  

## How was this patch tested?

With a unit test ("SPARK-26002: correct day of year calculations for Julian calendar years") which focuses to these corner cases.

Manually:

```
scala> sql("select year('1500-01-01')").show()

+------------------------------+
|year(CAST(1500-01-01 AS DATE))|
+------------------------------+
|                          1500|
+------------------------------+

scala> sql("select dayOfYear('1100-01-01')").show()

+-----------------------------------+
|dayofyear(CAST(1100-01-01 AS DATE))|
+-----------------------------------+
|                                  1|
+-----------------------------------+
```

### CODE REVIEW

**Comment by attilapiros** [comment]
ping @10110346 @gatorsmile

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 1]
```diff
@@ -410,6 +410,30 @@ class DateTimeUtilsSuite extends SparkFunSuite {
...
+      // even the multiples of 100 as before 1582-10-4 the Julian calendar leap year calculation
+      // is used in which every multiples of 4 are leap years
+      c.set(year, 2, 1, 0, 0, 0)
+      assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 61)
+      assert(getMonth(getInUTCDays(c.getTimeInMillis)) === 3)
+
+      // For non-leap years:
+      c.set(year + 1, 2, 1, 0, 0, 0)
+      assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 60)
+    }
```
Comment (squito) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: this is good, but I think its worth adding checks for a couple of special cases:

* 1582-10-3
* 1582-10-14 (though I guess the meaning of "dayInYear" is not so clear in this case)
* 1600-01-01
* 1600-03-01

I think they'll all be OK after your change, but good to have a check.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 2]
```diff
@@ -410,6 +410,30 @@ class DateTimeUtilsSuite extends SparkFunSuite {
     assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 78)
   }
 
+  test("SPARK-26002: correct day of year calculations for Julian calendar years") {
+    TimeZone.setDefault(TimeZoneUTC)
```
[the remark this answers is not in the document]
Comment (attilapiros) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: Thanks! Good catch! It is actually not needed any more.

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala (hunk 1 above)
Comment (attilapiros) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: The last two (1600-01-01 and 1600-03-01) are already tested as 1600 is included in `(1000 to 1600 by 100)`.

I have added a new check for 1582-10-03. 

But I would not add an assert for 1582-10-14 without knowing that is really the correct value.

I have checked PostgreSQL but there this 10 days gap is not handled at all: from `SELECT EXTRACT(DOY FROM TIMESTAMP '1582-10-03 00:00:00');`  to `SELECT EXTRACT(DOY FROM TIMESTAMP '1582-10-16 00:00:00');` from day by day it is consecutive days from 276 to 289.

**Comment by squito** [comment]
lgtm

**Comment by dongjoon-hyun** [comment]
Retest this please.

**Comment by dongjoon-hyun** [comment]
cc @cloud-fan , @gatorsmile , @mgaido91

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 3]
```diff
@@ -410,6 +410,32 @@ class DateTimeUtilsSuite extends SparkFunSuite {
     assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 78)
   }
 
+  test("SPARK-26002: correct day of year calculations for Julian calendar years") {
```
Comment (cloud-fan) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: can we have an end-to-end test?

**Comment by cloud-fan** [comment]
what was the behavior before this PR?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala (hunk 3 above)
Comment (attilapiros) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: @cloud-fan 
Do you mean adding a new test to [datetime.sql](https://github.com/apache/spark/blob/421673a4f6ae8820b57bb1fb92d9fe03dc646cde/sql/core/src/test/resources/sql-tests/inputs/datetime.sql)? 

What about adding the ones I used as a manual testing (from the PR description)?

Comment (cloud-fan) [reply sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: SGTM

**Comment by attilapiros** [comment]
@cloud-fan It was off by one (at least at 1500, with each 100 years back in time the difference increases) as I described in the [Jira](https://issues.apache.org/jira/browse/SPARK-26002):

```
scala> sql("select dayOfYear('1500-01-02')").show()
+-----------------------------------+
|dayofyear(CAST(1500-01-02 AS DATE))|
+-----------------------------------+
|                                                           1|
+-----------------------------------+

scala> sql("select year('1500-01-01')").show()
+------------------------------+
|year(CAST(1500-01-01 AS DATE))|
+------------------------------+
|                                            1499|
+------------------------------+


scala> sql("select month('1500-01-01')").show()
+-------------------------------+
|month(CAST(1500-01-01 AS DATE))|
+-------------------------------+
|                                                   12|
+-------------------------------+


scala> sql("select dayOfYear('1500-01-01')").show()
+-----------------------------------+
|dayofyear(CAST(1500-01-01 AS DATE))|
+-----------------------------------+
|                                                      365|
+-----------------------------------+

```

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 4]
```diff
@@ -53,14 +53,27 @@ object DateTimeUtils {
 
   final val MILLIS_PER_DAY = SECONDS_PER_DAY * 1000L
 
-  // number of days in 400 years
+  // number of days in 400 years by Gregorian calendar
   final val daysIn400Years: Int = 146097
+
+  // In Julian calendar every year which is a multiple of 4 is a leap year.
+  // So in 400 years there are 3 extra leap years compared to the Gregorian calendar:
+  // 100, 200 and 300 (as multiples of 100 but not 400).
```
Comment (viirya) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: Actually I must read https://en.wikipedia.org/wiki/Julian_calendar#Year_length;_leap_years to understand why there are three extra days here. I'm not sure if this added comments are enough clear to others.

Comment (mgaido91) [reply sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: +1, some kind of rewording of:
```
The Gregorian calendar has the same months and month lengths as the Julian calendar, but, in the Gregorian calendar, year numbers evenly divisible by 100 are not leap years, except that those evenly divisible by 400 remain leap years.
```
would be great here IMHO

File: sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala [hunk 5]
```diff
@@ -585,20 +598,30 @@ object DateTimeUtils {
    * Return the number of days since the start of 400 year period.
    * The second year of a 400 year period (year 1) starts on day 365.
    */
-  private[this] def yearBoundary(year: Int): Int = {
-    year * 365 + ((year / 4 ) - (year / 100) + (year / 400))
+  private[this] def yearBoundary(year: Int, isGregorian: Boolean): Int = {
```
Comment (mgaido91) [inline sql/catalyst/src/main/scala/org/apache/spark/sql/catalyst/util/DateTimeUtils.scala]: shall we update also the `isLeapYear` method above?

File: sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala [hunk 6]
```diff
@@ -410,6 +410,32 @@ class DateTimeUtilsSuite extends SparkFunSuite {
...
+      assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 61)
+      assert(getMonth(getInUTCDays(c.getTimeInMillis)) === 3)
+
+      // For non-leap years:
+      c.set(year + 1, 2, 1, 0, 0, 0)
+      assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 60)
+    }
+
+    c.set(1582, 2, 1, 0, 0, 0)
+    assert(getDayInYear(getInUTCDays(c.getTimeInMillis)) === 60)
```
Comment (mgaido91) [inline sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/util/DateTimeUtilsSuite.scala]: can we also add a test for valid dates? Checking leap days, eg. 29-02-400 or 29-02-1600 or 29-02-1800, ...

**Comment by gatorsmile** [comment]
cc @MaxGekk who is touching this

**Comment by MaxGekk** [comment]
@gatorsmile Thanx for the ping. Actually current implementation support [Proleptic Gregorian calendar](https://en.wikipedia.org/wiki/Proleptic_Gregorian_calendar) even if it is not documented explicitly. Recently we switched on [this chronology](https://github.com/apache/spark/pull/23358/files#diff-3e4b85f54d2e75f1b5b845279749024eR38) for parsing/formatting date/timestamps  in JSON/CSV datasource (see #23150 and #23196) and in functions (see #23358).

If we support the hybrid calendar ([Julian](https://en.wikipedia.org/wiki/Julian_calendar) + [Gregorian](https://en.wikipedia.org/wiki/Gregorian_calendar)) here, it will introduce additional inconsistency. 

@attilapiros If you need the hybrid calendar for some reasons, you should extend the `Year` expression by additional argument to set chronology (calendar system) otherwise you can break user's apps potentially.

**Review by MaxGekk — COMMENTED** [review]
Just in case, SQL standard counts dates exclusively in the Gregorian calendar, even for years before that calendar was in use.

**Comment by attilapiros** [comment]
@MaxGekk thanks for the explanation, but I still a bit uncertain.  

My main intention was about correcting the *dayOfYear/month/year* operators as the result in the following example are very strange (and not consistent with PostgreSQL too):
```
scala> sql("select year('1500-01-01')").show()
+------------------------------+
|year(CAST(1500-01-01 AS DATE))|
+------------------------------+
|                          1499|
+------------------------------+

scala> sql("select month('1500-01-01')").show()
+-------------------------------+
|month(CAST(1500-01-01 AS DATE))|
+-------------------------------+
|                             12|
+-------------------------------+


scala> sql("select dayOfYear('1500-01-01')").show()
+-----------------------------------+
|dayofyear(CAST(1500-01-01 AS DATE))|
+-----------------------------------+
|                                365|
+-----------------------------------+
```

Running the corresponding expressions in PostgreSQL 9.6:
```
>    SELECT DATE_PART('year', TIMESTAMP '1500-01-01 0:0:0') as "year";

| year |
| ---- |
| 1500 |

>   SELECT DATE_PART('month', TIMESTAMP '1500-01-01 0:0:0') as "month";

| month |
| ----- |
| 1     |

>  SELECT DATE_PART('doy', TIMESTAMP '1500-01-01 0:0:0') as "doy";

| doy |
| --- |
| 1   |
```
 [View on DB Fiddle](https://www.db-fiddle.com/f/4rLEmF3xJjZqSaqA3h74Ky/1)

Is there any other DBMS I can take as a reference?

**Pull request closed without merging**